File: ios/Classes/trtc/AITranscriberManagerHandler.swift
//
//  AITranscriberManagerHandler.swift
//  tencent_rtc_sdk
//

import Flutter
import Foundation
import TXLiteAVSDK_Professional

class AITranscriberManagerHandler: NSObject, AITranscriberListener {
    private let channel: FlutterMethodChannel
    private var transcriberManager: AITranscriberManager?
    private var isListenerAdded = false
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }
    
    func release() {
        if let manager = transcriberManager, isListenerAdded {
            manager.remove(self)
            isListenerAdded = false
        }
        transcriberManager = nil
    }
    
    private func getTranscriberManager() -> AITranscriberManager? {
        if transcriberManager == nil {
            transcriberManager = TRTCCloud.sharedInstance().getAITranscriberManager()
        }
        return transcriberManager
    }
    
    private func ensureListenerAdded() {
        if !isListenerAdded {
            if let manager = getTranscriberManager() {
                manager.add(self)
                isListenerAdded = true
            }
        }
    }
    
    func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        switch call.method {
        case "startRealtimeTranscriber":
            startRealtimeTranscriber(call, result: result)
            return true
        case "stopRealtimeTranscriber":
            stopRealtimeTranscriber(call, result: result)
            return true
        case "pauseReceivingTranscriberMessage":
            pauseReceivingTranscriberMessage(call, result: result)
            return true
        case "resumeReceivingTranscriberMessage":
            resumeReceivingTranscriberMessage(call, result: result)
            return true
        case "addTranscriberListener":
            addTranscriberListener(call, result: result)
            return true
        case "removeTranscriberListener":
            removeTranscriberListener(call, result: result)
            return true
        default:
            return false
        }
    }
    
    private func startRealtimeTranscriber(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        ensureListenerAdded()
        
        let params = TranscriberParams()
        
        if let transcriberRobotId = Utils.getParamByKeyCanBeNull(call: call, param: "transcriberRobotId", result:
            result) as? String {
            params.transcriberRobotId = transcriberRobotId
        }
        
        if let sourceLanguage = Utils.getParamByKeyCanBeNull(call: call, param: "sourceLanguage", result:
            result) as? String {
            params.sourceLanguage = sourceLanguage
        }
        
        if let userIds = Utils.getParamByKeyCanBeNull(call: call, param: "userIdsToTranscribe", result:
            result) as? [String] {
            params.userIdsToTranscribe = userIds
        }
        
        if let translationLanguages = Utils.getParamByKeyCanBeNull(call: call, param: "translationLanguages", result:
            result) as? [String] {
            params.translationLanguages = translationLanguages
        }
        
        getTranscriberManager()?.startRealtimeTranscriber(params)
        result(nil)
    }
    
    private func stopRealtimeTranscriber(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let transcriberRobotId = Utils.getParamByKeyCanBeNull(call: call, param: "transcriberRobotId", result:
            result) as? String ?? ""
        getTranscriberManager()?.stopRealtimeTranscriber(transcriberRobotId)
        result(nil)
    }
    
    private func pauseReceivingTranscriberMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        getTranscriberManager()?.pauseReceivingMessage()
        result(nil)
    }
    
    private func resumeReceivingTranscriberMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        getTranscriberManager()?.resumeReceivingMessage()
        result(nil)
    }
    
    private func addTranscriberListener(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        ensureListenerAdded()
        result(nil)
    }
    
    private func removeTranscriberListener(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let manager = getTranscriberManager(), isListenerAdded {
            manager.remove(self)
            isListenerAdded = false
        }
        result(nil)
    }
    
    // MARK: - AITranscriberListener
    
    func onRealtimeTranscriberStarted(_ roomId: String, transcriberRobotId: String) {
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod("onRealtimeTranscriberStarted", arguments: [
                "roomId": roomId,
                "transcriberRobotId": transcriberRobotId,
            ])
        }
    }
    
    func onReceiveTranscriberMessage(_ roomId: String, message: TranscriberMessage) {
        DispatchQueue.main.async { [weak self] in
            let messageMap: [String: Any] = [
                "segmentId": message.segmentId ?? "",
                "speakerUserId": message.speakerUserId ?? "",
                "sourceText": message.sourceText ?? "",
                "translationTexts": message.translationTexts ?? [:],
                "timestamp": message.timestamp,
                "isCompleted": message.isCompleted,
            ]
            
            self?.channel.invokeMethod("onReceiveTranscriberMessage", arguments: [
                "roomId": roomId,
                "message": messageMap,
            ])
        }
    }
    
    func onRealtimeTranscriberStopped(_ roomId: String, transcriberRobotId: String, reason: Int) {
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod("onRealtimeTranscriberStopped", arguments: [
                "roomId": roomId,
                "transcriberRobotId": transcriberRobotId,
                "reason": reason,
            ])
        }
    }
    
    func onRealtimeTranscriberError(_ roomId: String, transcriberRobotId: String, error: Int, errorInfo: String) {
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod("onRealtimeTranscriberError", arguments: [
                "roomId": roomId,
                "transcriberRobotId": transcriberRobotId,
                "error": error,
                "errorInfo": errorInfo,
            ])
        }
    }
}

File: android/src/main/java/com/tencent/trtcplugin/trtc/AITranscriberManagerHandler.java
package com.tencent.trtcplugin.trtc;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.tencent.liteav.transcriber.AITranscriberManager;
import com.tencent.liteav.transcriber.AITranscriberManager.TranscriberParams;
import com.tencent.liteav.transcriber.AITranscriberManager.TranscriberMessage;
import com.tencent.liteav.transcriber.AITranscriberManager.AITranscriberListener;
import com.tencent.trtc.TRTCCloud;
import com.tencent.trtcplugin.utils.MethodCallParams;
import com.tencent.trtcplugin.utils.TRTCLogger;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class AITranscriberManagerHandler implements AITranscriberListener {
    private static final String TAG = "AITranscriberManagerHandler";

    private Context mContext;
    private MethodChannel mChannel;
    private Handler mMainHandler = new Handler(Looper.getMainLooper());

    private AITranscriberManager mTranscriberManager;
    private boolean mIsListenerAdded = false;

    public AITranscriberManagerHandler(Context context, MethodChannel channel) {
        mContext = context;
        mChannel = channel;
    }

    public void release() {
        if (mTranscriberManager != null && mIsListenerAdded) {
            mTranscriberManager.removeListener(this);
            mIsListenerAdded = false;
        }
        mTranscriberManager = null;
        mMainHandler.removeCallbacksAndMessages(null);
    }

    private AITranscriberManager getTranscriberManager() {
        if (mTranscriberManager == null) {
            TRTCCloud trtcCloud = TRTCCloud.sharedInstance(mContext);
            mTranscriberManager = trtcCloud.getAITranscriberManager();
        }
        return mTranscriberManager;
    }

    private void ensureListenerAdded() {
        if (!mIsListenerAdded) {
            AITranscriberManager manager = getTranscriberManager();
            if (manager != null) {
                manager.addListener(this);
                mIsListenerAdded = true;
            }
        }
    }

    public boolean handleMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        try {
            Method method = AITranscriberManagerHandler.class.getDeclaredMethod(call.method, MethodCall.class, MethodChannel.Result.class);  // CHECKSTYLE:SUPPRESS LineLength
            method.invoke(this, call, result);
            return true;
        } catch (NoSuchMethodException e) {
            return false;
        } catch (Exception e) {
            TRTCLogger.e(TAG + " | method=" + call.method + " | error=" + e);
            return false;
        }
    }

    @SuppressWarnings("unused")
    private void startRealtimeTranscriber(MethodCall call, MethodChannel.Result result) {
        ensureListenerAdded();

        TranscriberParams params = new TranscriberParams();
        params.transcriberRobotId = MethodCallParams.getParamCanBeNull(call, result, "transcriberRobotId");
        params.sourceLanguage = MethodCallParams.getParamCanBeNull(call, result, "sourceLanguage");

        List<String> userIds = MethodCallParams.getParamCanBeNull(call, result, "userIdsToTranscribe");
        if (userIds != null) {
            params.userIdsToTranscribe = userIds;
        }

        List<String> translationLanguages = MethodCallParams.getParamCanBeNull(call, result, "translationLanguages");
        if (translationLanguages != null) {
            params.translationLanguages = translationLanguages;
        }

        AITranscriberManager manager = getTranscriberManager();
        if (manager != null) {
            manager.startRealtimeTranscriber(params);
        }
        result.success(null);
    }

    @SuppressWarnings("unused")
    private void stopRealtimeTranscriber(MethodCall call, MethodChannel.Result result) {
        String transcriberRobotId = MethodCallParams.getParamCanBeNull(call, result, "transcriberRobotId");

        AITranscriberManager manager = getTranscriberManager();
        if (manager != null) {
            manager.stopRealtimeTranscriber(transcriberRobotId != null ? transcriberRobotId : "");
        }
        result.success(null);
    }

    @SuppressWarnings("unused")
    private void pauseReceivingTranscriberMessage(MethodCall call, MethodChannel.Result result) {
        AITranscriberManager manager = getTranscriberManager();
        if (manager != null) {
            manager.pauseReceivingMessage();
        }
        result.success(null);
    }

    @SuppressWarnings("unused")
    private void resumeReceivingTranscriberMessage(MethodCall call, MethodChannel.Result result) {
        AITranscriberManager manager = getTranscriberManager();
        if (manager != null) {
            manager.resumeReceivingMessage();
        }
        result.success(null);
    }

    @SuppressWarnings("unused")
    private void addTranscriberListener(MethodCall call, MethodChannel.Result result) {
        ensureListenerAdded();
        result.success(null);
    }

    @SuppressWarnings("unused")
    private void removeTranscriberListener(MethodCall call, MethodChannel.Result result) {
        AITranscriberManager manager = getTranscriberManager();
        if (manager != null && mIsListenerAdded) {
            manager.removeListener(this);
            mIsListenerAdded = false;
        }
        result.success(null);
    }

    @Override
    public void onRealtimeTranscriberStarted(String roomId, String transcriberRobotId) {
        mMainHandler.post(() -> {
            Map<String, Object> params = new HashMap<>();
            params.put("roomId", roomId);
            params.put("transcriberRobotId", transcriberRobotId);
            mChannel.invokeMethod("onRealtimeTranscriberStarted", params);
        });
    }

    @Override
    public void onReceiveTranscriberMessage(String roomId, TranscriberMessage message) {
        mMainHandler.post(() -> {
            Map<String, Object> params = new HashMap<>();
            params.put("roomId", roomId);

            Map<String, Object> messageMap = new HashMap<>();
            messageMap.put("segmentId", message.segmentId);
            messageMap.put("speakerUserId", message.speakerUserId);
            messageMap.put("sourceText", message.sourceText);
            messageMap.put("translationTexts", message.translationTexts != null ? message.translationTexts : new HashMap<>());  // CHECKSTYLE:SUPPRESS LineLength
            messageMap.put("timestamp", message.timestamp);
            messageMap.put("isCompleted", message.isCompleted);

            params.put("message", messageMap);
            mChannel.invokeMethod("onReceiveTranscriberMessage", params);
        });
    }

    @Override
    public void onRealtimeTranscriberStopped(String roomId, String transcriberRobotId, int reason) {
        mMainHandler.post(() -> {
            Map<String, Object> params = new HashMap<>();
            params.put("roomId", roomId);
            params.put("transcriberRobotId", transcriberRobotId);
            params.put("reason", reason);
            mChannel.invokeMethod("onRealtimeTranscriberStopped", params);
        });
    }

    @Override
    public void onRealtimeTranscriberError(String roomId, String transcriberRobotId, int error, String errorInfo) {
        mMainHandler.post(() -> {
            Map<String, Object> params = new HashMap<>();
            params.put("roomId", roomId);
            params.put("transcriberRobotId", transcriberRobotId);
            params.put("error", error);
            params.put("errorInfo", errorInfo);
            mChannel.invokeMethod("onRealtimeTranscriberError", params);
        });
    }
}
